src/event_stream.rs

```diff
@@ -45,26 +42,34 @@ impl<S> EventStream for S
     where S: EventStepper
 {
     type Item = <S as EventStepper>::Item;
-    type Error = TryRecvError;
 
-    fn next_ev(&mut self) -> Result<Self::Item, Self::Error> {
+    fn next_ev(&mut self) -> Result<Self::Item, RecvError> {
         if let Some(cached_ev) = self.pop_item() {
             return Ok(cached_ev);
         }
-        if let Ok(new_events) = self.produce_events() {
-            self.buffer_items(new_events);
+        match self.produce_events() {
+            Ok(new_events) => {
+                self.buffer_items(new_events);
+                // Recursing here either returns one of the buffered events we just added,
+                // or in the case where `new_events` was empty, causes us to block again waiting
+                // on `produce_events`.
+                self.next_ev()
```

Comment:
I don't think Rust guarantees to do tail call elimination; couldn't this cause a stack overflow, depending on the implementation of `produce_events`? Might be safer to use a loop.